### ISSUE TITLE
No USB monitors detected after upgrading to newest version

### ISSUE THREAD

**Issue by zim2323** [terrymacdonald/DisplayMagician#193]
**Describe the bug**
DM no longer detects USB displays after upgrading from 2.4.1.14 to 2.5.0.352.

**To Reproduce**
Steps to reproduce the behaviour:

**Expected behaviour**
I have 6 monitors.  3 on RTX 4090, and 3 on J5create 4xHDMI USB-C adapter.  I should see all 6 monitors in DM.  I do see them in Display Settings.

**Screenshots**
Screenshot attached.

**Reporting Logs**
If DisplayMagician starts:
- Open DisplayMagician and perform whatever actions are needed to make the bug occur.
- Click on the Settings button on the main DisplayMagician window
- Click on 'Create a Support Zip File' button
- Save the Support Zip File on your computer.
- Come back here, and upload the Support Zip File through your web browser and attach it to this issue.

If DisplayMagician does not:
- Open a terminal window and run `C:\Program Files\DisplayMagician\DisplayMagician.exe --trace` to create a TRACE-level DisplayMagician.log file.
- The above command will create a DisplayMagician.log file in `C:\Users\<yourusername>\AppData\Local\DisplayMagician\Logs` which should contain a LOT of detailed TRACE level log entries. 
- Come back here, and upload the DisplayMagician.log file through your web browser and attach it to this issue.
- Also attach the following files to this issue:
  - `C:\Users\<yourusername>\AppData\Local\DisplayMagician\Profiles\DisplayProfiles_2.1.json` 
  - `C:\Users\<yourusername>\AppData\Local\DisplayMagician\Shortcuts\Shortcuts_2.0.json`
  - `C:\Users\<yourusername>\AppData\Local\DisplayMagician\Settings_2.0.json`

**Enviroment (please complete the following information):**
 - Windows Version: [e.g. Win10]  Windows 10 64bit Enterprise 22H2
 - DisplayMagician Version [e.g. 1.1.0]  2.5.0.352
 - Video Card: [e.g. Asus NVIDIA GTX1070]  Nvidia FE 4090
 - Number and make/model of monitors: [e.g. 2xDell UH2718H, 1x LG G27U17D] 3 AOC 32" (Nvidia), 3 AOC 27" (USB)
 - Date of last video driver update: [e.g. 2021-02-15]  2/8/2023
 - Date of last windows update: [e.g. 2021-02-15]  2/4/2023

**Additional context**
![example](https://user-images.githubusercontent.com/8365728/217699763-5dadb731-fbc8-4478-88f5-d8ce80434a65.jpg)
[DisplayMagician-Support-20230208-2117.zip](https://github.com/terrymacdonald/DisplayMagician/files/10692859/DisplayMagician-Support-20230208-2117.zip)

**Comment by terrymacdonald** [comment]
Hi,

Does changing between display profiles still work? Do all six screens still
get positioned correctly?

In other words, is it just the picture in the Display Profiles list that is
wrong?

Thanks

*Terry *

**Comment by zim2323** [comment]
It looks to just be the picture.  I was able to swap from surround and back
to individual displays without issues related to the top monitors.  I have
not tried turning the USB monitors off to see if it will turn them back on
yet.

**Comment by terrymacdonald** [comment]
ok - give that a go (disabling the USB monitors in Windows Display
Settings). Save that as a new profile and see if it will work for you.

At this stage I expect it's just the drawing part that's buggy. I had to
change the code as it wasn't properly working with some other video cards,
and it's entirely possible I broke the code for USB monitors. I don't own a
USB monitor so no chance for me to test :).

Can you please generate a Support ZIP File for me please and upload it to
the Github issue through your web browser? That will let me see the errors
in more detail. To create a Support ZIP File please:

   - Start DisplayMagician
   - Click on the *Settings *button
   - Check that the *Log Level *is set to TRACE (if not, change it to TRACE)
   - Go to the Display Profile window and try to swap between a couple of
   profiles (which will hopefully generate some errors in the logs)
   - Go back to the DisplayMagician Settings page
   - Click on the *Create Support ZIP File *button.
   - Save the ZIP File to your PC
   - Upload the ZIP File to this Github issue *through your web browser*.


Thanks

*Terry *

**Comment by terrymacdonald** [comment]
@zim2323 it looks like your SupportZIP file was created from DisplayMagician v2.4.1.14. I'll need you to install DisplayMagician 2.5.0 and try again using that, as that is the latest codebase that I'm working with. I won't be able to release a hotfix for v2.4.1.14 as I've updated the code now.

Thanks
Terry

**Comment by zim2323** [comment]
I'll do some more testing when I get home this evening.

Interesting that those logs show the old version.  I know for sure I took
those logs after the upgrade, and the version on the window definitely is
2.5.  I've pulled new logs and will upload for comparison.

**Comment by zim2323** [comment]
[DisplayMagician-Support-20230209-0830.zip](https://github.com/terrymacdonald/DisplayMagician/files/10697662/DisplayMagician-Support-20230209-0830.zip)

**Comment by terrymacdonald** [comment]
Hi @zim2323,

Looks like the DM version is still the old 2.4.1.14: 
```
2023-02-01 08:36:06.9448|INFO|DisplayMagician.Program|Program/Main: Starting DisplayMagician v2.4.1.14|
```
That version had a few little bugs in it, so we really do need you to upgrade. The upgrade process may not be working, so you may need to manually install v2.5.0.352 from here: https://github.com/terrymacdonald/DisplayMagician/releases/tag/v2.5.0

You can uninstall the older version of DisplayMagician first if you really want to be sure, as DisplayMagician keeps the old config files around, and they should be picked up automatically by the reinstalled v2.5.0.352.

Once the upgrade has happened please then firstly change the logging level in Settings to TRACE.  Once that is done please try and see if you can create a brand new display profile with the 3 J5create 4xHDMI USB-C adapter connected monitors. If you hit any issues please create a new Support ZIP File and upload here to this issue **using your web browser** (email attachments don't work).

Thanks
Terry

**Comment by zim2323** [comment]
Uninstalled, reinstalled, rebooted.   Deleted current profile and recreated.  Still shows only 3 monitors on GPU.  
[DisplayMagician-Support-20230309-1848.zip](https://github.com/terrymacdonald/DisplayMagician/files/10937209/DisplayMagician-Support-20230309-1848.zip)

**Comment by zim2323** [comment]
Sorry, I meant any word on this issue.

**Comment by terrymacdonald** [comment]
Hi @zim2323,

Um... I'm not sure how to say this, but your install is still showing 2.4.1.14. 

```
2023-02-01 08:36:06.9448|INFO|DisplayMagician.Program|Program/Main: Starting DisplayMagician v2.4.1.14|
```

Looking at the logged time shown above, I **think** that I know what's the issue here. That is a really old log date in the file. 

I think your new DisplayMagician 2.5.1 must have been installed in a different location to standard "C:\Program Files\DisplayMagician\", and the config files are not in the standard "C:\Users\<username>\AppData\Local\DisplayMagician". This would mean that your new DisplayMagician is logging to a different file somewhere else, meaning that those files that you send me are the old files from your previous DM install (which logged to the standard location).

So, can you please try and find where your DisplayMagician is installed on your PC, and where your new DisplayMagician is logging to? I will need to see those new files in order to troubleshoot your install.

When you find the location for the DisplayMagician config files, please don't use the Create Support ZIP File button to create the zip package (it will always look in the default config file location). Please manually create a ZIP file of the whole DisplayMagician config structure. It will have a structure that looks similar to this:

![image](https://user-images.githubusercontent.com/5721772/227806546-64b4ade2-1c3d-40f7-aa07-d2fcb6d157dc.png)

Zip up the entire directory structure and upload it to this issue with your web browser, and hopefully we'll finally get to the point that I can troubleshoot this :).

NOTE: Make sure your logging level is set to TRACE within DisplayMagician settings window so I get all the logs I need.

Thanks
Terry

**Comment by terrymacdonald** [comment]
Hi @zim2323 

Have you had any chance to manually zip up the logs for me so I can do some more troubleshooting?

Thanks
Terry

**Comment by zim2323** [comment]
Sorry, I just finished up a massive PC upgrade and haven't had a chance to
track those down yet.  I did start to look prior and I don't see anything
out of the ordinary.  My Son just started soccer for his school and another
club so my weekdays will be pretty tight.  I'm hoping to have some time
this weekend to look at it again.

Thanks!
Chris

**Comment by zim2323** [comment]
I apologize for taking so long to get back to this.  Life tends to get in the way of things when you have a family.  I've completely uninstalled, reinstalled, and grabbed both the program created ZIP as well as a manual ZIP of the directory specified.
[DisplayMagician-Support-20230925-1947.zip](https://github.com/terrymacdonald/DisplayMagician/files/12720575/DisplayMagician-Support-20230925-1947.zip)
[DisplayMagician.zip](https://github.com/terrymacdonald/DisplayMagician/files/12720577/DisplayMagician.zip)

**Comment by zim2323** [comment]
Update...

1)  I completely removed 2.5.0.352, rebooted and loaded 2.4.1.14 fresh with a final reboot, thinking I'd go back to what works until you have a chance to diagnose the problem.  When I did this, DM would crash after clicking out of the "upgrade/info" screen.  I tried running as Administrator, anything I could do, but would not stay running.  I upgraded to 2.5.0.352, rebooted, and everything is working without the USB displays showing.

2)  I got a chance to save and switch between the 6 monitors in standard(6 separate displays) and surround+3 successfully.  So, even though the displays don't show up in the DM GUI, they are in fact included.  I "think" I can confirm this is the case because if i just switch from 6 to surround+3 from Nvidia Control Panel, the 3 top (USB) displays are disabled until I manually enable and place them.  By saving and swapping profiles in DM the displays return to their exact configuration as when I saved the DM Profile.

So "MAYBE" this is all just a graphical issue?

Thanks!
Chris

**Comment by zim2323** [comment]
Yup.  Everything works.  Just what gets displayed.  Ive purposely changed
screens not shown and they work as expected.

**Comment by terrymacdonald** [comment]
@zim2323  I think this issue was related to a bug I recently found in the screen processing logic. I have completely rewritten the logic in this area, and I will be shortly releasing v2.6.0. This should hopefully fix your issue. I will post a link in this issue for you to download a test version of the software. Once you can confirm this fixes things then I'll be hopefully able to release this to the general public.

Expect something in the next 2 weeks.

Thanks
Terry

**Comment by terrymacdonald** [comment]
@zim2323  I've created a new test version of DisplayMagician that *may* fix the issue that you're experiencing, and I've attached it to this Github issue. Please download the attached zip file, extract the msi installer from the zip file, and then run the installer. It should upgrade your existing Display Profiles.

[DisplayMagicianSetup-v2.6.0.446.zip](https://github.com/terrymacdonald/DisplayMagician/files/14018624/DisplayMagicianSetup-v2.6.0.446.zip)

This new version of DisplayMagician has the ability to 'update' Display Profiles - effectively overwriting the display layout with the one that is currently being used. So if you start If you still start having issues with a Display Profile not working, simply:
- setup your display layout the way you want it
- then open DisplayMagician's Display Profiles Window
- select on the saved Display Profile you wish to update
- click the Update button
- say 'Yes' when asked if you really want to overwrite settings

Your DisplayMagician Display Profile will now be recreated using the current display settings, and they will overwrite the old settings. This will hopefully get rid of any display layout issues, and it doesn't require any changes to your Game Shortcuts. Nice!

There is a single known issue at the moment, which is that the NVIDIALibrary SetDisplayConfig function doesn't work for NVIDIA video cards. Luckily this issue shouldn't have much impact at all as the WindowsLibrary functions should do enough configuration of their own to ensure it all works. If you have some special NVIDIA settings that you use then this version of DisplayMagician *may* not apply them correctly. You will unfortunately just have to test if that is the case for your particular setup. 

I am currently troubleshooting the NVIDIALibrary SetDisplayConfig function error but it looks like it could be a bit of a pain to fix, so I didn't want to hold up the test release; I'd rather get the test version into your hands so I can see what other bugs are in there!

If this fixes your issue please reply here, and if you find another bug please log a new Github issue so I can track them separately.

Thanks
Terry

**Comment by zim2323** [comment]
Thanks Terry!

New version installed and it is working now.  I see all 6 monitors.  I'll
let you know if I run into any issues.

I had to do an uninstall and remove old
c:\users\<user>\appdata\local\displaymagician folder and then reinstall in
order for it to work.  Just doing an in-place upgrade displaymagician.exe
immediately stopped.


Chris

**Comment by terrymacdonald** [comment]
@zim2323 Thanks for that confirmation about it needing to replace the config. That's also been logged under #283 so I'll start working on the problem under that. This is sort of error is so useful to know, so thank you for testing this for me. It means I can fix it before doing a main public release!

Glad to hear that you can see all the monitors now. There is still a potential for issues as I've changed the way that DisplayMagician works. It used to decide on a mode (e.g. NVIDIA mode) and then only use that mode. Now it uses all video libraries and tries to configure things in this order: NVIDIA -> AMD -> Windows. In my testing on multiple different machines this worked, but with over 8,000 current installs of DisplayMagician all on different hardware and combinations of hardware I'm sure there will be some issues uncovered.

Thanks
Terry

**Comment by zim2323** [comment]
Glad to hear Terry!

Is it possible to include an override in the settings to let the
user choose which library to use?  It could be a thing if you have more
than one type of video card installed in your system.

Thanks,
Chris

**Comment by terrymacdonald** [comment]
@zim2323  I did find some issues with the earlier v2.6.0.446 version. I've made a lot of revisions to how DisplayMagician works in v2.6.0.485 (attached) that should make it way more reliable. But it is highly likely you will need to 'update' your Display Profiles to make them work properly.

I'd recommend first of all applying a Display Profile, and then press the 'update' button. That will recreate the config for the Display Profile and should make it work properly. You'll need to do that to each of your saved Display Profiles. Then please test out changing between your Display Profiles to see if everything works as expected.

Here is the latest test version of DisplayMagician: 
[DisplayMagicianSetup-v2.6.0.485.zip](https://github.com/terrymacdonald/DisplayMagician/files/14075432/DisplayMagicianSetup-v2.6.0.485.zip)

Thanks for all your help testing.

Terry

**Comment by terrymacdonald** [comment]
Thank you for this. This confirms to me that the AMD library needs a better
way of detecting when there isn't an AMD chip present.

Can you please confirm if you have an AMD CPU with a built in GPU? Or do
you run an Intel chipset? Does it have an Intel GPU in it? Do you run
NVIDIA GPUs?

Thanks
Terry

**Comment by zim2323** [comment]
I run an Intel i9 13900KS and an Nvidia 4090.  No AMD video in the
system at all.  The ASUS ROG board doesn't seem to have any AMD chips
onboard either.  No Intel GPU.

Chris

**Comment by terrymacdonald** [comment]
Perfect thanks. That confirms the issue.

I'll see what I can work on over the next few days to improve the AMD
library party of DisplayMagician.

This will also be a problem for AMD only PC's as the NVIDIALibrary will
also have the same issue (the reverse of yours).

Now I know the issue I should be able to find a fix that works for both
groups of users, and hopefully DM goes back to being nice and reliable!

Thanks for all your testing, it really, really helps me a lot.

Thanks
Terry

**Comment by zim2323** [comment]
I should correct my last email.  I do have built in Intel processor GPU.
No drivers installed.

Would still be nice to have settings to disable specific libraries.  Would
be great troubleshooting option.

Chris

**Comment by terrymacdonald** [comment]
Unfortunately with the new logic it's impossible to do this. And the new
logic is required to fix a very large number of users with gaming laptops
that have AMD and NVIDIA GPUs together. DisplayMagician want working for
them at all.

An alternative is to use the NVIDIAInfo, AMDInfo and CCDInfo command line
applications that I've also written. They are designed to be used
separately, and are what I actually test each specific library with.

There is also VideoInfo, which tests them all combined together, but that
one doesn't fully work at the moment, hence the need to perform tests using
DisplayMagician itself.

Thanks
Terry

**Comment by zim2323** [comment]
Understand what you're saying.  Just to clarify, I'm not talking about a
bullet selection.  I'm talking checkboxes, with the ability to
check/uncheck any number of libraries to attempt to use.  This would allow
laptop users with multiple chipsets to still select and function.

That said, I understand the complications and I'll wait for the next
version to test.

Chris

**Comment by terrymacdonald** [comment]
@zim2323, I've added in more checks and logic to avoid crashes. Could you please try this latest version? 
[DisplayMagicianSetup-v2.6.0.508.zip](https://github.com/terrymacdonald/DisplayMagician/files/14135107/DisplayMagicianSetup-v2.6.0.508.zip)

I'm hoping this will work, but at the very least it should allow DisplayMagician to boot up enough for you to create a Support ZIP File. Let's see whether I've fixed it 🤞.

**Comment by terrymacdonald** [comment]
@zim2323 The latest test version should be working properly, and hopefully fixes your issue. Could you please test it for me so I know that it fixes your issue? If it does then I'll be able to close this issue as the fix will be part of the next full release.

[DisplayMagicianSetup-V2.6.0.515-No-NvApi_SetDisplayConfig.zip](https://github.com/terrymacdonald/DisplayMagician/files/14200905/DisplayMagicianSetup-V2.6.0.515-No-NvApi_SetDisplayConfig.zip)

As I've not really had any feedback from you recently I'll give it a week and then close this issue if I have no further responses. We can always reopen this issue in the future if need be.

Thanks!

**Comment by terrymacdonald** [comment]
Oops I broke the setup with my previous version.  This version has a corrected setup that should install everything properly.

[DisplayMagicianSetup-v2.6.0.522-corrected-setup.zip](https://github.com/terrymacdonald/DisplayMagician/files/14249946/DisplayMagicianSetup-v2.6.0.522-corrected-setup.zip)

Thanks
Terry

**Comment by zim2323** [comment]
I apologize.  Work has been crazy and i have had no time to test yet.
Hopefully this week.

Chris

**Comment by whittedb** [comment]
@terrymacdonald, I just found your application after deciding to try some NVIDIA surround for gaming with my 6 monitor setup.  Of course, the first thing I see is this issue and here I am.  It looks like @zim2323 is pretty busy at the moment.  I have a bit more time on my hands and would be willing to help test this to get it working.

I'm running 3x27" monitors hanging off an RTX 2070 and 3xUSB monitors.  I created and deleted a single profile in an attempt to figure out why I wasn't seeing all 6 in the graphic screen.  Should I grab that last install you posted or do you have another one ready that has new changes?

**Comment by zim2323** [comment]
Morning Terry,

Finally had a chance to install and test latest beta.  So far things look
good.  I'm able to save profiles now.  I'll need to do some further testing
but so far so good.  I'll let you know if I run into anything.

Thanks!
Chris Miller

**Comment by whittedb** [comment]
I tried that last install you supplied and it seems to work for me now too.  Quick OT question... Is there a way to configure and save different NVIDIA Surround setups?  I'd like to have one triple monitor surround and 2 double monitor surround (2 left and 2 right) that I can switch between.

**Comment by zim2323** [comment]
Brad,
I've done similar.  Just make changes you want, go back into DM and save
the new Profile.  Do this for each layout.  Once saved you can swap back
and forth.

**Comment by whittedb** [comment]
@zim2323, That didn't work for me.  It seems to always use the last NVIDIA surround setup I had configured.

**Comment by zim2323** [comment]
I haven't gone through and rebuilt all my profiles again yet.  I still need
to do my additional Surround profiles, so I'll let you know what I find
out.  Each profile should be specific to whatever setup you are wanting to
use.

**Comment by terrymacdonald** [comment]
Hi Brad,

DisplayMagician grabs the Surround Display Layout (Grid layout) from the
NVIDIA driver if it is in use when you save a new Display Profile (or
'update' a Display Profile). It *should *be able to handle creating
different Surround Display Profiles as it creates the Display Grids itself
each time and then gives them to the NVIDIA driver to use. This is all done
programmatically through the NvAPI.

There may still be some limitations within the NVIDIA driver that might
prevent this from working in all cases. NVIDIA has an internal database of
settings that it uses for tracking all the different layouts that users
create, and it may be that the way that DisplayMagician operates may
conflict with the settings in the NVIDIA driver sometimes. There isn't
anything I can do about that, as it is just the way that NVIDIA has decided
to do things.

If you ever find you have a problem with a Surround Display Profile, I
suggest recreating your Surround display layout the way you want it to be,
then Alt-Tab back to DisplayMagician and click on the 'update' button to
get DisplayMagician to grab the current settings from the NVIDIA and
Windows drivers. This effectively refreshes your Display Profile and should
clear out any NVIDIA settings that don't make sense any longer.

Thanks
Terry

**Comment by whittedb** [comment]
Ok guys, thanks for the input.  I'll play around with this again.  Maybe I was just tired that night and missed a step.

**Comment by terrymacdonald** [comment]
Hi!

This may be fixed by the brand new version of DisplayMagician I just released. I would suggest that you try to the new DisplayMagician v2.7.0-alpha.1 test build available here: 
 https://github.com/terrymacdonald/DisplayMagician/releases/tag/v2.7.0-alpha.1

It should fix your problem, but it may also cause other problems instead. This version has had a complete backend change, new installer, new features, and so it may have some bugs that need squashing. 

Please test it out. and if you find any new problems please log a new issue. If it fixes this issue then please update this issue to tell me and then close this issue (it helps me keep track of what I've still got to fix).

Thanks
Terry

**Comment by terrymacdonald** [comment]
@whittedb Are you able to test the version I just released please? It now has the ability to 'Update' a Display Profile which is much friendlier.

When you get an issue with a display layout., you simply recreate the layout using your NVIDIA setup/Windows Display Settings, and then swap back to DisplayMagician. You then select the Display Profile you want to update, and then click the Update button. That will replace your previous Display Profile settings with the ones you have now, and then all your Game Shortcuts will automatically use the newly updated settings. 

It's WAAAAAAY easier than it was before.

Hope that helps.
Terry

**Comment by shayan229** [comment]
Hi, I have had the same issue as the user who opened this ticket. I can confirm that my monitor connected via a dell usb c dock is now detected and functions as other monitors do with the alpha build 2.7.0.40069.
![Screenshot 2024-10-04 173746](https://github.com/user-attachments/assets/fff6b6ef-25c6-49bd-9f48-ebeb58a3eae9)
For reference the monitor labelled "WINDOWS Display" is he one connected via usb and was not detected before.

**Comment by terrymacdonald** [comment]
Oh excellent! Thanks for the update.

Terry

**Comment by terrymacdonald** [comment]
Ok, I'm going to close this issue, but feel free to reopen it if you get the same bug occurring.
